sklearn_numba_dpex/common/kernels.py

```diff
@@ -228,31 +235,57 @@ def make_sum_reduction_2d_axis1_kernel(
 
     Notes
     -----
-    `work_group_size` is assumed to be a power of 2.
+    if `size1` is None then the kernel expects 1d tensor inputs, and `axis` parameter
+    is ignored.
 
-    if `size1` is None then the kernel expects 1d tensor inputs. If `size1` is not None
-    then the expected shape of input tensors is `(size0, size1)`, and the reduction
-    operation is equivalent to input.sum(axis=1). Depending on the size of the sum
-    axis, it might be worth tuning `work_group_size` for better performance.
+    If `size1` is not None then the expected shape of input tensors is
+    `(size0, size1)`, and the reduction operation is equivalent to
+    `input.sum(axis=axis)`.
+
+    if `size1` is None or if `axis` is `0`, then `work_group_size` is assumed to be a
+    power of 2, and the parameter `sub_group_size` is ignored.
+
+    if `size1` is not `None` and `axis` is `1`, then `work_group_size` is assumed to be
+    a multiple of `sub_group_size`, such that `work_group_size // sub_group_size` is
+    a power of 2.
+
+    Depending on the size of the sum axis, it might be worth tuning `work_group_size`
+    for better performance. When `axis` is `1`, tuning `sub_group_size` with respect
+    to the size of the other axis might also be beneficial.
     """
-    if size1 is not None:
-        n_rows = size0
+    if is_1d := not size1:
+        axis = 1
+        size1 = size0
+        size0 = 1
+
+    if axis == 1:
+        work_group_size, kernels = _make_partial_sum_reduction_2d_axis1_kernels(
+            size0, work_group_size, fused_unary_func, dtype, device
+        )
         sum_axis_size = size1
+        get_result_shape = lambda result_sum_axis_size: (size0, result_sum_axis_size)
+        get_global_size = (
+            lambda result_sum_axis_size: result_sum_axis_size * work_group_size * size0
```

Comment:
Why do we not use this?
```suggestion
            lambda result_sum_axis_size: result_sum_axis_size 
            * work_group_size 
            * math.ceil(size0 / sub_group_size)
```

Reply:
When axis=1, each work group perform a local reduction on axis 1 in a window of size (1, 2 * work_group_size). (does not depend on sub_group_size)

When axis=0, each work group perform a local reduction on axis 0 in a window of size (sub_group_size_,work_group_size // sub_group_size).

That's why the global size and the size of the intermediate results are different in both cases.

And the reason for those differences is that it's done to optimize the IO which I think is the main bottleneck, and the IO pattern is different in both cases, so are the kernels.

I added some more comments explaining that.